torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,290 @@
+      result = xm.all_gather(input_tensor, groups=self._mesh)
+      for i, slice in enumerate(torch.split(result, input_tensor.shape[0])):
+        output_tensors[i].copy_(slice)
+
+    return WorkXla([t for sublist in output_tensors_list for t in sublist])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/release/1.10/torch/distributed/distributed_c10d.py#L1129
+  def broadcast(self, tensors, opts):
+    root_tensor = tensors[opts.rootTensor]
+    root_rank = opts.rootRank
+    if root_rank != self.rank():
+      with torch.no_grad():
+        root_tensor.zero_()
+    xm.all_reduce(xm.REDUCE_SUM, [root_tensor], groups=self._mesh)
+
+    return WorkXla([root_tensor])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/release/1.10/torch/distributed/distributed_c10d.py#L2355
+  def reduce_scatter(self, output_tensors, input_tensors_list, opts):
+    for input_tensors, output_tensor in zip(input_tensors_list, output_tensors):
+      # Ensure all inputs have the same shape.
+      first_shape = input_tensors[0].shape
+      for i, t in enumerate(input_tensors[1:]):
+        if first_shape != t.shape:
+          raise ValueError(f"Input {i+1}'s shape is different from input 0: "
+                           f"{t.shape} vs {first_shape}")
+      input_tensor = torch.cat(input_tensors)
+      reduce_type = self._get_reduce_type(opts.reduceOp)
+      groups = self._mesh
+      shard_count = len(groups[0]) if groups else self.size()
+      xm.reduce_scatter(
+          reduce_type,
+          input_tensor,
+          scatter_dim=0,
+          shard_count=shard_count,
+          scale=1,
+          groups=groups,
+          output=output_tensor)
+
+    return WorkXla(output_tensors)
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2683
+  def barrier(self, opts):
+    return WorkXla()
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L1417
+  # `reduce` is not needed by DeepSpeed for now.
+  def reduce(self, *args):
+    raise NotImplementedError
+
+  def allgather_coalesced(self, *args):
+    raise NotImplementedError
+
+  def allreduce_coalesced(self, *args):
+    raise NotImplementedError
+
+  def alltoall(self, *args):
+    raise NotImplementedError
+
+  def alltoall_base(self, *args):
+    raise NotImplementedError
+
+  def gather(self, *args):
+    raise NotImplementedError
+
+  def scatter(self, *args):
+    raise NotImplementedError
+
+  def send(self, *args):
+    raise NotImplementedError
+
+  def recv(self, *args):
+    raise NotImplementedError
+
+  def recv_anysource(self, *args):
+    raise NotImplementedError
+
+  def monitored_barrier(self, *args):
+    raise NotImplementedError
+
+  def Options(self, *args):
+    raise NotImplementedError
+
+
+class WorkXla(Work):
```

Comment:
nit. should comment/describe the class better, as it's not obvious from the name.

Reply:
We should fix that in the master.